### PR TITLE
IDGEN-42 web-services for IDGEN

### PR DESCRIPTION
**Pull request by brayoh**
## JIRA TICKET NAME:
[IDGEN-42: Web-services for idgen](https://issues.openmrs.org/projects/IDGEN/issues/IDGEN-42)

## SUMMARY:
Merge the implementation at [openmrs-module-idgen-webservices](https://github.com/Bhamni/openmrs-module-idgen-webservices) back to idgen.

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/idgen/contract/IdentifierSource.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/idgen/contract/IdentifierSource.java]: Can you change the license to what we are currently using?

File: api/src/main/java/org/openmrs/module/idgen/contract/IdentifierType.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/idgen/contract/IdentifierType.java]: Can you change the license to what we are currently using?

File: api/src/main/java/org/openmrs/module/idgen/mapper/IdentifierSourceListMapper.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/idgen/mapper/IdentifierSourceListMapper.java]: Can you change the license to what we are currently using?

File: api/src/main/java/org/openmrs/module/idgen/serialization/ObjectMapperRepository.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/idgen/serialization/ObjectMapperRepository.java]: Can you change the license to what we are currently using?

File: api/src/main/java/org/openmrs/module/idgen/util/PrivilegeConstants.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/idgen/util/PrivilegeConstants.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/IdgenWsConstants.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
+ *
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/IdgenWsConstants.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/controller/GenerateIdentifierRequest.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/controller/GenerateIdentifierRequest.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierSourceController.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierSourceController.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierTypeController.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierTypeController.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/controller/SetLatestIdentifierRequest.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/controller/SetLatestIdentifierRequest.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapper.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapper.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapperImpl.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapperImpl.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierTypeServiceWrapper.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierTypeServiceWrapper.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierTypeServiceWrapperImpl.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierTypeServiceWrapperImpl.java]: Can you change the license to what we are currently using?

File: omod/src/test/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierSourceControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierSourceControllerTest.java]: Can you change the license to what we are currently using?

File: omod/src/test/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierTypeControllerTest.java [hunk 16]
```diff
@@ -0,0 +1,94 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierTypeControllerTest.java]: Can you change the license to what we are currently using?

File: omod/src/test/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierTypeResourceTest.java [hunk 17]
```diff
@@ -0,0 +1,87 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierTypeResourceTest.java]: Can you change the license to what we are currently using?

File: omod/src/test/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapperImplTest.java [hunk 18]
```diff
@@ -0,0 +1,213 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapperImplTest.java]: Can you change the license to what we are currently using?

File: omod/src/test/java/org/openmrs/module/idgen/web/services/IdentifierTypeServiceWrapperImplTest.java [hunk 19]
```diff
@@ -0,0 +1,119 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/services/IdentifierTypeServiceWrapperImplTest.java]: Can you change the license to what we are currently using?

File: omod/src/main/java/org/openmrs/module/idgen/web/IdgenWsConstants.java [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+	
+	/*
+	 * URIs
+	 */
+	public static final String PATH_IDGEN_IDTYPE = "idgen/identifiertype";
+	
+	/*
+	 * GPs
+	 */
+	public static final String GP_PRIMARY_IDTYPE = "bahmni.primaryIdentifierType";
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/IdgenWsConstants.java]: Can we replace bahmni with openmrs?

File: omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapperImpl.java [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+    public List<org.openmrs.module.idgen.contract.IdentifierSource> getAllIdentifierSources() {
+        IdentifierSourceService identifierSourceService = Context.getService(IdentifierSourceService.class);
+        List<IdentifierSource> allIdentifierSources = identifierSourceService.getAllIdentifierSources(false);
+        return IdentifierSourceListMapper.map(allIdentifierSources);
+    }
+
+    @Override
+    public List<org.openmrs.module.idgen.contract.IdentifierSource> getAllIdentifierSourcesOfPrimaryIdentifierType() {
+        IdentifierSourceService identifierSourceService = Context.getService(IdentifierSourceService.class);
+        String primaryIdentifierTypeUuid = Context.getAdministrationService().getGlobalProperty("bahmni.primaryIdentifierType");
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapperImpl.java]: Can you replace bahmni with openmrs?

File: omod/src/test/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierTypeControllerTest.java [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+        assertEquals(HttpStatus.UNAUTHORIZED, response.getStatusCode());
+    }
+
+    @Test
+    public void ShouldReturnAllIdentifierType() throws Exception {
+        when(Context.isAuthenticated()).thenReturn(true);
+        when(Context.hasPrivilege("Get Identifier Types")).thenReturn(true);
+
+        final IdentifierSource identifierSource = new IdentifierSource("source-uuid", "some name", "SOM");
+        final IdentifierType bahmniIdentifierType = new IdentifierType("uuid", "bahmni", "description", ".*", true, true, Arrays.asList(identifierSource));
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/controller/IdgenIdentifierTypeControllerTest.java]: All over here, can you replace bahmni with openmrs?

File: omod/src/test/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapperImplTest.java [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+        assertEquals("uuid",(allIdentifierSources.get(0).getUuid()));
+        assertEquals("GAN",(allIdentifierSources.get(0).getPrefix()));
+    }
+
+    @Test
+    public void shouldGetAllIdentifierSourcesBasedOnPrimaryIdentifierType() {
+        when(Context.getService(IdentifierSourceService.class)).thenReturn(identifierSourceService);
+
+        when(Context.getAdministrationService()).thenReturn(administrationService);
+        when(administrationService.getGlobalProperty("bahmni.primaryIdentifierType")).thenReturn("dead-cafe");
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/services/IdentifierSourceServiceWrapperImplTest.java]: Can you replace bahmni with openmrs?

File: omod/src/test/java/org/openmrs/module/idgen/web/services/IdentifierTypeServiceWrapperImplTest.java [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+    public void setUp() throws Exception {
+        initMocks(this);
+        mockStatic(Context.class);
+        serviceWrapperImp = new IdentifierTypeServiceWrapperImpl();
+
+    }
+
+    @Test
+    public void shouldGetAllIdentifierType() throws Exception {
+        final IdentifierType primaryIdentifierType = new IdentifierType("primary-identifier-type-uuid", "bahmni", "description", ".*", true, true, new ArrayList<IdentifierSource>());
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/services/IdentifierTypeServiceWrapperImplTest.java]: Can you replace bahmni with openmrs?

**Comment by brayoh** [comment]
@dkayiwa You mean I revert this changes and instead integrate the ones in the other PR?

**Comment by dkayiwa** [comment]
@brayoh yes. I thought you had read the JIRA ticket comments where it was advised as so.

**Comment by brayoh** [comment]
Let me go through the comments again.

**Pull request closed without merging**